### DISCUSSION TITLE
Project management

### DISCUSSION THREAD

**Discussion by ann0see** [jamulussoftware/jamulus#1131] [Documentation]
I think becoming more and more important to think about project management in general. Although we have the issues tab, a lot of issues are not currently being worked on. Although we have a release structure, I think, a clearer project management should be implemented. What are your thoughts?

**Comment by gene96817** [comment]
It will help organize issues into related milestones.

**Reply by ann0see** [reply]
Great! I will give you access in a few minutes.

**Reply by ann0see** [reply]
Done

**Comment by gilgongo** [comment]
One thing that came up in discussion was a "roadmap" of some kind. I've had a look around on other FOSS and similar projects and there's a lot of wide variation in the concepts. @gene96817 is that what you mean my "milestones"?  I was wondering whether we might have "Current phase", "Next Phase", "Future phases"? So a bit like Debian's Stable, Testing, Unstable conceptions, only we'd not have builds, just feature priorities, and not date driven.

**Reply by gene96817** [reply]
Yes, this would be a good way to organize new ideas and collections of related changes.

**Reply by gilgongo** [reply]
OK well I'll see if I can comb through the GH issues and do a first pass at collecting them into those groups we can discuss if it makes sense. It's really only so that we have a kind of official way of setting what we think are priorities so that we don't spend too much time discussing some of the more exotic feature requests that come in (particularly if they're made by people who can't implement them I guess).

**Reply by gilgongo** [reply]
OK I've imported the current list of GH issues from the main repo here (editing is on for all):

https://docs.google.com/spreadsheets/d/1EVq-VS5nbctWwe8K5xx4j8eAISM1LxldJqOV-xDIkgY/edit?usp=sharing

There a lot... I might start by looking for ones that we might flag for migration to a discussion? The goal being to get to an issues list that is for reasonably firmed up features, or bugs so that we can then create Epics and Stories and assign to Phases/Milestones(?). 

BTW quite how we do this depends on the outcome of the conversation we're having about discussion platforms though.

Feel free to put anything in the "Actions" column.

BTW if anyone else wants to create their own, I got a CSV using the "hub" tool. Clone and cd into the repo and do:

`hub issue -s all -f "%t|%l|%S|%au|%NC|%cD|%uD|%I|%U%n" > issues.csv`

**Reply by gilgongo** [reply]
Looking in more detail, there's only really a handful that are clearly agreed feature requests or bugs. The rest seem to be discussions. Also worth considering whether we might in fact combine code and docs tickets into one place, while having separate code and docs discussions. That way, we can have all the agreed "to dos" in one place, and code and docs are pretty closely coupled anyway (and I expect the docs to stabilise over time so that updates to them will be a lot less frequent than now).

**Reply by pljones** [reply]
I'm tempted to agree:
- documentation bug: this means the documentation doesn't document the existing (correct) application behaviour
- application bug: this means the documentation is correct but the application behaviour doesn't match (or is wrong and not documented, of course)
- any other documentation change should only happen with a change in application behaviour and be _part of_ that change, to ensure it's reviewed along with the application change

So "documentation bug" is just another type of bug, rather than something completely separate.

**Reply by gilgongo** [reply]
OK Peter and I have done a first pass through the list.

- We could start with closing a bunch, so I can put comments on them with an explanation and do that.
- There's quite few that are legit but the status is unclear so I could ping authors etc. on those to ask.
- That leaves the quite large number of "discussion" ones, ~~which depends on [what we want to use](https://github.com/jamulussoftware/jamuluswebsite/discussions/303#discussioncomment-373875) for that...~~ which we can move to the relevant GitHub Discussions now we're up for that.

Meanwhile there are a few we see differently I think:

1. UDP fragmentation seems to need more discussion before it can be an agreed work ticket? https://github.com/jamulussoftware/jamulus/issues/299 
3. OS notifications seems pretty clear cut to me, does it need more discussion? https://github.com/jamulussoftware/jamulus/issues/425
5. Re. closing https://github.com/jamulussoftware/jamulus/issues/497 but if somebody implemented it, would there be a problem?

PS: The fact that Server performance & optimization https://github.com/jamulussoftware/jamulus/issues/455 has 287 comments on it is a bit of a LOL.

**Reply by pljones** [reply]
> UDP fragmentation seems to need more discussion before it can be an agreed work ticket?

I think solving the problem needs to be a priority.  I don't know of a solution, though.  So yes, more work is needed before a clear solution can be written up - as far as I'm aware.

> OS notifications seems pretty clear cut to me, does it need more discussion?

Hm.  Maybe I was over-thinking the problem at the time.  If you don't consider anything but mixer channels, you could always alert if the mixer changed.  So there would need to be a setting to enable this for those who can see such alerts but not the mixer.  (I was thinking of having the server push a notification when someone joined or left - but the effect is that the mixer changes at the client, of course, so the client can work it out.)

> Re. closing jamulussoftware/jamulus#497 but if somebody implemented it, would there be a problem?

For people who didn't want it, there would be.  So it would need to be a server UI and command line option, plus the support to pass that into the recorder start up.  Again, I've not heard it being a problem for many people, so I'd expect the default to be UTC as it is now.  My experience with time zones is that they make code obscure and hard to maintain...

**Reply by pljones** [reply]
> Server performance & optimization

That was raised as a discussion before Github had them, I think.  It also spawned about three or four branches off master, IIRC...  It's not "finished"..!  (It doesn't have any completion criteria...)

**Reply by gilgongo** [reply]
OK will record that in the sheet. 

Oh BTW once we're down to a solid list of actionable stuff in the issues (that is, fairly clear things to go and work on) I was going to ask: what do we think of the idea of adopting a convention of **editing** the "base" ticket in response to comments on it? 

This would be so that if someone wanted to pick something up, they'd have a decent summary of the current thinking available without having to read through all the comments to work it out.

**Reply by gilgongo** [reply]
BTW also today I'll start the process of closing out and moving various issues to discussion.

**Reply by gilgongo** [reply]
OK done that - mostly. 

I ended up taking a fairly conservative approach and left some pending a bit more info from the posters - in some cases it looked like they might be on the verge of being able to spec them up. 

Also BTW while I know that some are obviously in the "never going to happen" category, I thought it best to move them to 
 discussions rather than simply closing. If somebody wants to pick up a total re-write of the app to provide an input volume  controller or something would we not let them put in a PR?

Also interesting to see that the act of moving things to discussion has brought back some from the dead. The metronome redux!

This then leaves most things in Issues as at least reasonably stable descriptions but in many cases seemingly dormant. Might need to do a second pass over those now.

**Comment by gene96817** [comment]
One reason I suggested this process was that it appeared to me that we were addressing a lot of ideas as single stand-alone concepts. Putting some of the requests into related bundles could yield many benefits and efficiencies.

**Reply by pljones** [reply]
"Epics" and "Stories" if we want to play SAFe.

**Reply by gilgongo** [reply]
I rashly thought I'd try combing thought the issues to see if I can a) have a go at suggesting some road map groupings, and b) have a go at suggesting Epic/Story sub-groupings. Probably doing this in spreadsheet I think. Would that be a good idea or is there a better way of doing it?

**Reply by gene96817** [reply]
That is a good start. A spreadsheet would be a lightweight tool. 
{I would like to see all the scenarios that involve managing participants (aka admission /attendance control) grouped together.}

**Reply by pljones** [reply]
We've reverted to spreadsheets at work having tried a few other things, both for planning and maintaining the projects in flight.  It's a "higher level" tool: less detail - but then, that detail can obscure the message, sometimes.  So it depends what you're trying to show.

I wouldn't suggest full-blown CPAN diagram... (not used those since my final year at secondary school...).

**Reply by gene96817** [reply]
I also have "reverted" to spreadsheets because it is much more nimble. I use grouped (nested?) rows or columns to carry details.

**Reply by gilgongo** [reply]
OK I'll see if I can compile a Google sheet of the issues and we can look at that. It could also be useful to help identify some general pruning and purging actions at some point so that we can get the Github Issues into a functional list and less of a bunch of blurry discussions masquerading a feature requests :-)

**Reply by ann0see** [reply]
Great! If you want me to help out, I'll help.

**Reply by gilgongo** [reply]
Sheet's here https://github.com/jamulussoftware/jamuluswebsite/discussions/273#discussioncomment-377351

**Reply by gene96817** [reply]
@gilgongo   Would you like to take a pass at grouping any of the rows by internal impact to the code? Maybe this is adding new column? Are there some obvious rows that are related by similar use cases?

**Reply by gilgongo** [reply]
The first main grouping that sticks out is the UI-related stuff (as Peter has remarked in fact - although I think it's mostly for discussion in fact). There may be some other related stuff too I guess...

**Comment by ann0see** [comment]
Moved this to the main repo since it's not only related to the documentation.

**Comment by gilgongo** [comment]
I think https://github.com/jamulussoftware/jamulus/discussions/929#discussioncomment-411918  has now taken over on this conversation so locking.
